File: src/Sems.Api/Modules/Analytics/Application/AnalyticsService.cs
using Sems.Api.Modules.Analytics.Domain.Model;
using Sems.Api.Modules.Analytics.Domain.Repositories;
using Sems.Api.Modules.Analytics.Domain.Services;
using Sems.Api.Shared.Errors;

namespace Sems.Api.Modules.Analytics.Application;

/// <summary>
/// Casos de uso del modulo de analitica.
///
/// <para>Comandos y consultas conviven aqui porque el contexto es pequenio:
/// cinco agregados con operaciones de crear, listar y cerrar. Separarlos en dos
/// servicios anadiria ceremonia sin aportar claridad.</para>
/// </summary>
public sealed class AnalyticsService
{
    private readonly IBillPredictionRepository _predictions;
    private readonly IRecommendationRepository _recommendations;
    private readonly IAnomalyRepository _anomalies;
    private readonly IDeviceIdentificationRepository _identifications;
    private readonly IConsumptionRankingRepository _rankings;
    private readonly IBillCalculator _calculator;

    public AnalyticsService(IBillPredictionRepository predictions,
        IRecommendationRepository recommendations, IAnomalyRepository anomalies,
        IDeviceIdentificationRepository identifications, IConsumptionRankingRepository rankings,
        IBillCalculator calculator)
    {
        _predictions = predictions;
        _recommendations = recommendations;
        _anomalies = anomalies;
        _identifications = identifications;
        _rankings = rankings;
        _calculator = calculator;
    }

    // -------------------------------------------------- proyeccion de recibo

    public Task<BillPrediction> CreatePredictionAsync(string userId, int year, int month,
        DateTime periodStart, DateTime periodEnd, double estimatedKwh, double estimatedAmount,
        string? currency, double tariffUsed, double errorMargin, CancellationToken ct = default) =>
        _predictions.SaveAsync(BillPrediction.Create(userId, year, month, periodStart, periodEnd,
            estimatedKwh, estimatedAmount, currency, tariffUsed, errorMargin), ct);

    public Task<List<BillPrediction>> PredictionsByUserAsync(string userId, CancellationToken ct = default) =>
        _predictions.FindByUserIdAsync(userId, ct);

    /// <summary>
    /// Calcula y guarda la proyeccion de recibo de un local con la tarifa
    /// comercial vigente.
    /// </summary>
    /// <remarks>
    /// A diferencia de <see cref="CreatePredictionAsync"/>, aqui el importe no
    /// lo trae quien llama: se calcula. Es lo que permite que la proyeccion
    /// refleje el cargo por potencia, que en un local puede ser la mitad de la
    /// factura y con la tarifa plana anterior no aparecia por ningun lado.
    /// </remarks>
    public Task<BillPrediction> ForecastSiteBillAsync(string userId, string siteId,
        string? tariffCategory, double contractedPowerKw, int year, int month,
        DateTime periodStart, DateTime periodEnd, double kwhPeak, double kwhOffPeak,
        double maxDemandKw, double errorMargin, CancellationToken ct = default)
    {
        var estimado = _calculator.Estimate(tariffCategory, kwhPeak, kwhOffPeak, maxDemandKw,
            contractedPowerKw);

        var prediccion = BillPrediction.Create(userId, year, month, periodStart, periodEnd,
            estimatedKwh: kwhPeak + kwhOffPeak,
            estimatedAmount: estimado.Total,
            currency: estimado.Currency,
            tariffUsed: estimado.TariffUsed,
            errorMargin: errorMargin,
            siteId: siteId,
            kwhPeak: kwhPeak,
            kwhOffPeak: kwhOffPeak,
            maxDemandKw: maxDemandKw,
            energyCost: estimado.EnergyCost,
            powerCost: estimado.PowerCost);

        return _predictions.SaveAsync(prediccion, ct);
    }

    // -------------------------------------------------------- recomendaciones

    public Task<Recommendation> CreateRecommendationAsync(string userId, string? deviceId,
        string? type, string? title, string? description, double savingKwh, double savingAmount,
        string? currency, CancellationToken ct = default) =>
        _recommendations.SaveAsync(Recommendation.Create(userId, deviceId, type, title, description,
            savingKwh, savingAmount, currency), ct);

    public Task<List<Recommendation>> RecommendationsByUserAsync(string userId,
        CancellationToken ct = default) => _recommendations.FindByUserIdAsync(userId, ct);

    public async Task<Recommendation> ApplyRecommendationAsync(Guid id, CancellationToken ct = default)
    {
        var recommendation = await _recommendations.FindByIdAsync(id, ct)
                             ?? throw AppException.NotFound($"Recommendation '{id}' not found");
        recommendation.Apply();
        return await _recommendations.SaveAsync(recommendation, ct);
    }

    // ------------------------------------------------------------- anomalias

    public Task<Anomaly> CreateAnomalyAsync(string userId, string? deviceId, string? type,
        string? description, string? severity, double actualKwh, double expectedKwh,
        CancellationToken ct = default) =>
        _anomalies.SaveAsync(Anomaly.Detect(userId, deviceId, type, description, severity,
            actualKwh, expectedKwh), ct);

    public Task<List<Anomaly>> AnomaliesByUserAsync(string userId, CancellationToken ct = default) =>
        _anomalies.FindByUserIdAsync(userId, ct);

    public async Task<Anomaly> ResolveAnomalyAsync(Guid id, CancellationToken ct = default)
    {
        var anomaly = await _anomalies.FindByIdAsync(id, ct)
                      ?? throw AppException.NotFound($"Anomaly '{id}' not found");
        anomaly.Resolve();
        return await _anomalies.SaveAsync(anomaly, ct);
    }

    // ---------------------------------------- identificacion de aparatos

    public Task<DeviceIdentificationResult> CreateIdentificationAsync(string userId,
        string? deviceId, string? predictedType, double confidence, string? status,
        CancellationToken ct = default) =>
        _identifications.SaveAsync(DeviceIdentificationResult.Create(userId, deviceId,
            predictedType, confidence, status), ct);

    public Task<List<DeviceIdentificationResult>> IdentificationsByUserAsync(string userId,
        CancellationToken ct = default) => _identifications.FindByUserIdAsync(userId, ct);

    // ----------------------------------------------------------------- ranking

    public Task<ConsumptionRanking> CreateRankingAsync(string userId, string? periodType,
        DateTime periodStart, DateTime periodEnd, IEnumerable<RankingItem>? items,
        CancellationToken ct = default) =>
        _rankings.SaveAsync(ConsumptionRanking.Create(userId, periodType, periodStart, periodEnd,
            items), ct);

    public Task<List<ConsumptionRanking>> RankingsByUserAsync(string userId,
        CancellationToken ct = default) => _rankings.FindByUserIdAsync(userId, ct);
}

File: src/main/java/com/sems/analytics/application/AnalyticsService.java
package com.sems.analytics.application;

import com.sems.analytics.domain.model.entities.*;
import com.sems.analytics.domain.model.valueobjects.RankingItem;
import com.sems.analytics.domain.repositories.AnalyticsRepositories.*;
import com.sems.shared.errors.AppException;
import java.time.Instant;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Casos de uso del modulo de analitica.
 *
 * <p>Comandos y consultas conviven aqui porque el contexto es pequenio: cinco
 * agregados con operaciones de crear, listar y cerrar. Separarlos en dos
 * servicios anadiria ceremonia sin aportar claridad.
 */
@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final BillPredictionRepository predictions;
    private final RecommendationRepository recommendations;
    private final AnomalyRepository anomalies;
    private final DeviceIdentificationRepository identifications;
    private final ConsumptionRankingRepository rankings;

    // ------------------------------------------------------ proyeccion de recibo

    @Transactional
    public BillPrediction createPrediction(String userId, int year, int month, Instant periodStart,
                                           Instant periodEnd, double estimatedKwh,
                                           double estimatedAmount, String currency,
                                           double tariffUsed, double errorMargin) {
        return predictions.save(BillPrediction.create(userId, year, month, periodStart, periodEnd,
                estimatedKwh, estimatedAmount, currency, tariffUsed, errorMargin));
    }

    @Transactional(readOnly = true)
    public List<BillPrediction> predictionsByUser(String userId) {
        return predictions.findByUserId(userId);
    }

    // ------------------------------------------------------------ recomendaciones

    @Transactional
    public Recommendation createRecommendation(String userId, String deviceId, String type,
                                               String title, String description, double savingKwh,
                                               double savingAmount, String currency) {
        return recommendations.save(Recommendation.create(userId, deviceId, type, title,
                description, savingKwh, savingAmount, currency));
    }

    @Transactional(readOnly = true)
    public List<Recommendation> recommendationsByUser(String userId) {
        return recommendations.findByUserId(userId);
    }

    @Transactional
    public Recommendation applyRecommendation(UUID id) {
        Recommendation recommendation = recommendations.findById(id)
                .orElseThrow(() -> AppException.notFound("Recommendation '" + id + "' not found"));
        recommendation.apply();
        return recommendations.save(recommendation);
    }

    // ------------------------------------------------------------------ anomalias

    @Transactional
    public Anomaly createAnomaly(String userId, String deviceId, String type, String description,
                                 String severity, double actualKwh, double expectedKwh) {
        return anomalies.save(Anomaly.detect(userId, deviceId, type, description, severity,
                actualKwh, expectedKwh));
    }

    @Transactional(readOnly = true)
    public List<Anomaly> anomaliesByUser(String userId) {
        return anomalies.findByUserId(userId);
    }

    @Transactional
    public Anomaly resolveAnomaly(UUID id) {
        Anomaly anomaly = anomalies.findById(id)
                .orElseThrow(() -> AppException.notFound("Anomaly '" + id + "' not found"));
        anomaly.resolve();
        return anomalies.save(anomaly);
    }

    // ------------------------------------------------- identificacion de aparatos

    @Transactional
    public DeviceIdentificationResult createIdentification(String userId, String deviceId,
                                                           String predictedType, double confidence,
                                                           String status) {
        return identifications.save(DeviceIdentificationResult.create(userId, deviceId,
                predictedType, confidence, status));
    }

    @Transactional(readOnly = true)
    public List<DeviceIdentificationResult> identificationsByUser(String userId) {
        return identifications.findByUserId(userId);
    }

    // -------------------------------------------------------------------- ranking

    @Transactional
    public ConsumptionRanking createRanking(String userId, String periodType, Instant periodStart,
                                            Instant periodEnd, List<RankingItem> items) {
        return rankings.save(ConsumptionRanking.create(userId, periodType, periodStart,
                periodEnd, items));
    }

    @Transactional(readOnly = true)
    public List<ConsumptionRanking> rankingsByUser(String userId) {
        return rankings.findByUserId(userId);
    }
}
